wallet/src/plan.rs

```diff
@@ -264,6 +268,272 @@ where
     Ok(plan)
 }
 
+#[allow(clippy::too_many_arguments)]
+#[allow(dead_code)]
+#[instrument(skip(fvk, view, rng, swap_nft_note, fee, source_address))]
+pub async fn swap_claim<V, R>(
+    fvk: &FullViewingKey,
+    view: &mut V,
+    mut rng: R,
+    swap_nft_note: Note,
+    fee: u64,
+    source_address: Option<u64>,
+) -> Result<TransactionPlan, anyhow::Error>
+where
+    V: ViewClient,
+    R: RngCore + CryptoRng,
+{
+    tracing::debug!(?swap_nft_note, ?fee, ?source_address);
+
+    return Err(anyhow::anyhow!("not implemented"));
+
+    let chain_params = view.chain_params().await?;
+
+    // TODO: need to fetch clearing price (`BatchSwapOutputData`) for the
+    // swap.
+
+    let mut plan = TransactionPlan {
+        chain_id: chain_params.chain_id,
+        fee: Fee(fee),
+        ..Default::default()
+    };
+
+    // Add a `SwapClaimPlan` action:
+    // plan.actions.push(
+    //     SwapClaimPlan::new(
+    //         &mut rng,
+    //         swap_nft_note,
+    //         swap_nft_position,
+    //         // The `fvk` is always the claim address.
+    //         fvk.address(),
+    //         fee,
+    //         output_data,
+    //         anchor,
+    //         trading_pair,
+    //     )
+    //     .into(),
+    // );
+
+    // The value we need to spend is 1 unit of the swap NFT, plus fees.
+    let mut value_to_spend: HashMap<Denom, u64> = HashMap::new();
+    // *value_to_spend.entry(swap_nft_note.denom()).or_default() += input_value;
+    // if fee > 0 {
+    //     *value_to_spend
+    //         .entry(STAKING_TOKEN_DENOM.clone())
+    //         .or_default() += fee;
+    // }
+
+    // Add the required spends:
+    for (denom, spend_amount) in value_to_spend {
+        if spend_amount == 0 {
+            continue;
+        }
+
+        let source_index: Option<AddressIndex> = source_address.map(Into::into);
+        // Select a list of notes that provides at least the required amount.
+        let notes_to_spend = view
+            .notes(NotesRequest {
+                fvk_hash: Some(fvk.hash().into()),
+                asset_id: Some(denom.id().into()),
+                address_index: source_index.map(Into::into),
+                amount_to_spend: spend_amount,
+                include_spent: false,
+            })
+            .await?;
+        if notes_to_spend.is_empty() {
+            // Shouldn't happen because the other side checks this, but just in case...
+            return Err(anyhow::anyhow!("not enough notes to spend",));
+        }
+
+        let change_address_index: u64 = fvk
```

Comment:
Ah right, for the SwapClaim, we spend a note with 1 NFT _plus_ we need to spend enough notes to cover fees, which can result in change. I wonder if there's a nice way to factor out this change logic since it's largely the same afaict with the `swap` function (could be done in a followup / filed as an issue).

Reply:
Hrm, this isn't quite right for `SwapClaim`, because the fee is actually pre-paid during the `Swap` and the `SwapClaim` doesn't require a spend authorization (but wait, how do you spend the swap NFT without a spend auth?). @hdevalence

Reply:
spending the swap NFT is "internalized" into the `SwapClaim` action, so there shouldn't be anything to do other than construct a `SwapClaimPlan`.